### PR TITLE
Add basic validation implementation

### PR DESCRIPTION
**Pull request by jonnydgreen**
Closes: https://github.com/mercurius-js/validation/issues/1

Draft PR that adds a basic implementation. At the moment, I have just focussed on proving the concept for external policies:
 - JSON Schema (ajv)
 - JTD (ajv)
 - Functional approach (field arguments only)

This has the following caveats (which I think are okay for v1?):

 - Functional approach only works for arguments. I think extending this beyond arguments is pretty complicated and maybe better for future versions?
 - I am only considering input data. Again, for v1, I think validating on outputs is too much for the time being and just focussing on inputs is probably a better initial approach?

Let me know what you think about the above caveats and what I've done so far. The quality of the code isn't great atm as there are a lot of things to polish up, but the core ideas are there if you'd like to take an early look :)

Still to do:

- [x] Directive support (using JSON Schema and external policy logic)
- [x] List support
- [x] Schema type inference
- [x] Better options validation
- [x] Make sure it works in the gateway
- [x] Deduplication of code and general clean up
- [x] Get coverage to 100%
- [x] Add ability to register custom formats using the Ajv API
- [x] Typings
- [x] Examples
- [x] Benchmarks
- [x] Documentation (inc. function docstrings)

Open Questions:

- Do we want to support validation of full types for v1?

### CODE REVIEW

File: test/jtd-validation-policy.js [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+    const app = Fastify()
+    t.teardown(app.close.bind(app))
+
+    app.register(mercurius, {
+      schema,
+      resolvers
+    })
+    app.register(mercuriusValidation, {
+      validationSchemaMode: 'JTD',
```
Comment (mcollina) [inline test/jtd-validation-policy.js:72]:
```suggestion
      mode: 'JTD',
```

File: test/jtd-validation-policy.js [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+      validation: {
```
Comment (mcollina) [inline test/jtd-validation-policy.js:73]:
```suggestion
      schema: {
```

File: test/json-schema-validation-policy.js [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+    const app = Fastify()
+    t.teardown(app.close.bind(app))
+
+    app.register(mercurius, {
+      schema,
+      resolvers
+    })
+    app.register(mercuriusValidation, {
+      validation: {
```
Comment (mcollina) [inline test/json-schema-validation-policy.js:74]:
```suggestion
      schema: {
```

**Review by mcollina — COMMENTED** [review]
Some bikeshedding on options name

**Comment by jonnydgreen** [comment]
Made some more progress - next steps are to tidy things up a bit, add extra testing and then move onto:
 - Documentation
 - Types
 - Examples

File: test/jtd-validation.js [hunk 4]
```diff
@@ -936,6 +900,102 @@ t.test('JTD validators', t => {
...
+      schema,
+      resolvers
+    })
+    app.register(mercuriusValidation, {
+      mode: 'JTD',
+      schema: {
+        Filters: {
+          __typeValidation: {
+            values: {
+              type: 'uint8'
```
Comment (mcollina) [inline test/jtd-validation.js:936]: how is this associated to the `id` field?

File: test/advanced-validation.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
+'use strict'
+
+const t = require('tap')
+const Fastify = require('fastify')
+const mercurius = require('mercurius')
+const mercuriusValidation = require('..')
+
+const schema = `
+  ${mercuriusValidation.graphQLTypeDefs}
```
[the remark this answers is not in the document]
Comment (jonnydgreen) [reply test/advanced-validation.js:9]: yep, it should be a 1-1 mapping as I build the typeDefs from an Ajv instance. Note, I've only included the keywords that have primitive values for v1, although I'm thinking of adding support for `items` etc (I personally have this use case!). You can view the supported keywords here: https://github.com/mercurius-js/validation/pull/2/files#diff-4ccc94c4b948f4abeb2e87f342463f374ba5502d926f387bd86e58bde6c028b0R23 :)

**Comment by jonnydgreen** [comment]
Almost there with the documentation. I'll push that up with the examples tonight and then crack on with the typings :)

File: test/jtd-validation.js (hunk 4 above)
Comment (mcollina) [reply test/jtd-validation.js:936]: Yes exactly

**Comment by mcollina** [comment]
Amazing work!

**Comment by jonnydgreen** [comment]
Thanks! It's been very fun to code :)

**Comment by jonnydgreen** [comment]
Finalising the rest of the docs after a couple of changes - once that's done this afternoon I'll bring it out of draft :)

File: docs/directive-validation.md [hunk 6]
```diff
@@ -0,0 +1,344 @@
+# Directive validation
+
+By default, Mercurius validation supports `@constraint` Directives out of the box. It is defined as follows:
```
Comment (mcollina) [inline docs/directive-validation.md:3]: Is there a way to disable this?

**Review by mcollina — COMMENTED** [review]
Good work, just a nit!

File: README.md [hunk 7]
```diff
@@ -1 +1,192 @@
...
+
+```
+
+### Gateway GraphQL Server Mode | With Validation
+
+Last run: `2021-09-18`
+
+```text
+
+```
```
[the remark this answers is not in the document]
Comment (jonnydgreen) [reply README.md:184]: Good spot! I'll re-run the benchmarks again now and fill in the missing slots :)

File: docs/directive-validation.md (hunk 6 above)
Comment (jonnydgreen) [reply docs/directive-validation.md:3]: Not at the moment, I can add an option to disable directive validation for sure! The total options would look like?

```js
app.register(mercuriusValidation, {
  mode: 'JTD' | 'JSONSchema', // Default 'JSONSchema'
  directiveValidation: false, // Default true
  schema: {
	...
  }
})
```

Comment (mcollina) [reply docs/directive-validation.md:3]: Yes exactly!

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**